cpp/src/arrow/flight/transport/grpc/grpc_server.cc

```diff
@@ -336,21 +406,35 @@ class GrpcServiceHandler final : public FlightService::Service {
       if (instance != nullptr) {
         flight_context.middleware_.push_back(instance);
         flight_context.middleware_map_.insert({factory.first, instance});
-        instance->SendingHeaders(&outgoing_headers);
       }
     }
 
+    // TODO factor this out after fixing all streaming and non-streaming handlers
+    if (!skip_headers) {
+      addMiddlewareHeaders(context, flight_context);
+    }
+
     return ::grpc::Status::OK;
   }
 
+  void addMiddlewareHeaders(ServerContext* context,
```

Comment:
it might be simpler to just have this be idempotent and then you can just sprinkle calls to it everywhere where we return to gRPC, and call it explicitly in a few places?

Reply:
Do we explicitly prohibit (I don't think so?!) threaded handler implementations?  I wanted to ensure that this is blocking so it works correctly there.  I did consider creating the blocking call wrapper object in the handlers instead and passing it to the stream wrapper c'tors, but only once I had poked the ugly hole back into the stream wrappers to handle the "RPC handler didn't actually write to the stream and trigger middleware header ingestion" case i.e. the `stream.TryCallOnceBeforeWrite()`.  If you think that one is too offensive I will find the time to relocate where the blocking call wrapper is scoped.

All of that said, I could (as you suggest?) roll the blocking call wrapper into AddMiddlewareHeaders but I wanted to avoid the locking dance for the non-streaming cases (not that it matters incredible much I guess).

Reply:
And all of that said, to truly fix this issue everywhere we do need to inject it as a hook into the stream wrappers so that it can be called between "most of" the Flight server's handler code and when response(s) are written to the stream—so not just everywhere we return to gRPC or anywhere we can call it explicitly.

Reply:
Threading is fine, but calls need to be synchronized (except Read and Write can be done concurrently - no other operations can be called concurrently)